storage/storage_image.go

```diff
@@ -66,6 +67,11 @@ type storageImageDestination struct {
 	filenames       map[digest.Digest]string        // Mapping from layer blobsums to names of files we used to hold them
 	SignatureSizes  []int                           `json:"signature-sizes,omitempty"`  // List of sizes of each signature slice
 	SignaturesSizes map[digest.Digest][]int         `json:"signatures-sizes,omitempty"` // Sizes of each manifest's signature slice
+	// fields to support early commit of layers below
+	indexLock         sync.Mutex              // Mutex to synchronize access to below fields
```

Comment:
Not `indexToStorageID`, it seems. Please make that clearer, similarly document what exactly `putBlobMutex` covers, and then review the users.

Reply:
Alternatively, do we even need two different locks if all we are protecting (IIRC) is short non-blocking data accesses?

Reply:
> Alternatively, do we even need two different locks if all we are protecting (IIRC) is short non-blocking data accesses?

Very good thinking. I'll investigate that idea.

Reply:
> Not indexToStorageID, it seems. Please make that clearer, similarly document what exactly putBlobMutex covers, and then review the users.

`indexToStorageID` is protected implicitly since only _one_ goroutine can be the "worker" which in turn is guaranteed via the `indexLock`.

Reply:
You’re right, `indexToStorageID` uses are safe. I meant that the comment “to synchronize access to below fields” was inaccurate WRT that field.

These things should be accurately documented for future readers of the code.
Or, speculatively, maybe something like
```go
type storageImageDestination struct {
    …
    locked struct {
        lock sync.Mutex
        … // THESE fields require the lock to be held, and it’s clearly visible at every use
    }
   …
}
```
? Obviously that works much better with languages with generics that can actually ~enforce that accesses only happen with a lock held, but it might be worth it even here just because comments go so easily out of sync with behavior.

Reply:
I will prepare a follow-up PR to improve the comments for future generations :) Thank you, as always, for your very constructive and thoughtful review!

Reply:
https://github.com/containers/image/pull/1188